### PR TITLE
Improving speed of Pytest Suite

### PR DESCRIPTION
**Pull request by devendragovil**
This PR improves the run times of the CI/CD pipeline bringing run times down from ~ 22 minutes to around 5.5 minutes. 

These are the approaches used:

1. Parallelization of the tests taking advantage of the 2 CPUs that Github Action runners have
2. Splitting pytests based on previous run times, to have even partition of tests
3. Reducing number of epochs in tutorials
4. Refactoring of the code and CI/CD pipeline

This largely addresses #215 

There are some techniques that can be used later for further speedups.

### CODE REVIEW

File: test/nn/simplicial/test_scnn.py [hunk 1]
```diff
@@ -38,7 +38,7 @@ def test_forward(self):
         incidence_1 = simplicial_complex.incidence_matrix(rank=1)
         incidence_2 = simplicial_complex.incidence_matrix(rank=2)
         laplacian_0 = simplicial_complex.hodge_laplacian_matrix(rank=0, weight=True)
-        laplacian_down_1 = simplicial_complex.down_laplacian_matrix(rank=1, weight=True)
+        laplacian_down_1 = simplicial_complex.down_laplacian_matrix(rank=1, weight=None)
         laplacian_up_1 = simplicial_complex.up_laplacian_matrix(rank=1)
         laplacian_2 = simplicial_complex.hodge_laplacian_matrix(rank=2, weight=True)
```
Comment (ffl096) [inline test/nn/simplicial/test_scnn.py:43]: See above.

File: .github/workflows/docs.yml [hunk 2]
```diff
@@ -2,7 +2,7 @@ name: "Docs: Check and Deploy"
 
 on:
   push:
-    branches: [main, github-actions-test]
+    branches: [main, github-actions-test,ghaction-speedup-iss215]
```
Comment (ffl096) [inline .github/workflows/docs.yml:5]: This should not come into the main branch, please remove from this pull request

File: .github/workflows/lint.yml [hunk 3]
```diff
@@ -2,7 +2,7 @@ name: "Lint"
 
 on:
   push:
-    branches: [main,github-actions-test]
+    branches: [main,github-actions-test,ghaction-speedup-iss215]
```
Comment (ffl096) [inline .github/workflows/lint.yml:5]: This should not come into the main branch, please remove from this pull request

File: .github/workflows/test_codebase.yml [hunk 4]
```diff
@@ -1,8 +1,9 @@
-name: "Test"
+name: "Testing Codebase"
 
-on:
+on: 
+  workflow_dispatch:
   push:
-    branches: [main,github-actions-test]
+    branches: [main,github-actions-test,ghaction-speedup-iss215]
```
Comment (ffl096) [inline .github/workflows/test_codebase.yml:6]: This should not come into the main branch, please remove from this pull request

File: test/nn/simplicial/test_scnn.py [hunk 5]
```diff
@@ -38,7 +38,7 @@ def test_forward(self):
         incidence_1 = simplicial_complex.incidence_matrix(rank=1)
         incidence_2 = simplicial_complex.incidence_matrix(rank=2)
         laplacian_0 = simplicial_complex.hodge_laplacian_matrix(rank=0, weight=True)
```
[the remark this answers is not in the document]
Comment (devendragovil) [reply test/nn/simplicial/test_scnn.py:40]: change to weight=None

File: test/nn/simplicial/test_sccnn.py [hunk 6]
```diff
@@ -43,7 +43,7 @@ def test_forward(self):
         incidence_1 = simplicial_complex.incidence_matrix(rank=1)
         incidence_2 = simplicial_complex.incidence_matrix(rank=2)
         laplacian_0 = simplicial_complex.hodge_laplacian_matrix(rank=0, weight=True)
```
[the remark this answers is not in the document]
Comment (devendragovil) [reply test/nn/simplicial/test_sccnn.py:45]: changed to weight=None

File: test/nn/simplicial/test_sccnn.py [hunk 7]
```diff
@@ -43,7 +43,7 @@ def test_forward(self):
...
-        laplacian_down_1 = simplicial_complex.down_laplacian_matrix(rank=1, weight=True)
+        laplacian_down_1 = simplicial_complex.down_laplacian_matrix(rank=1, weight=None)
         laplacian_up_1 = simplicial_complex.up_laplacian_matrix(rank=1)
         laplacian_2 = simplicial_complex.hodge_laplacian_matrix(rank=2, weight=True)
```
[the remark this answers is not in the document]
Comment (devendragovil) [reply test/nn/simplicial/test_sccnn.py:48]: Done

File: test/nn/simplicial/test_scnn.py (hunk 1 above)
Comment (devendragovil) [reply test/nn/simplicial/test_scnn.py:43]: Done

File: .github/workflows/docs.yml (hunk 2 above)
Comment (devendragovil) [reply .github/workflows/docs.yml]: I have made the change. Sorry for this being left over!

File: .github/workflows/lint.yml (hunk 3 above)
Comment (devendragovil) [reply .github/workflows/lint.yml]: I have made the change. Sorry for this being left over!

File: .github/workflows/test_codebase.yml (hunk 4 above)
Comment (devendragovil) [reply .github/workflows/test_codebase.yml]: I have made the change. Sorry for this being left over!

**Review by ffl096 — APPROVED** [review]

**Pull request merged**